충남대 BE_좌진우 Step3 - 위시 리스트

jinwoojwa opened the pull request:
## Step 3 - 위시 리스트

### 3단계 미션 구현 

✅ API 명세

| URL                                                   | 메서드    | 기능           | 설명                       |
|-------------------------------------------------------|--------|--------------|--------------------------|
| /api/wishes                                           | POST   | 위시 리스트 상품 추가 | 회원의 위시 리스트에 상품을 추가한다.    |
| /api/wishes/{wishId}                                  | DELETE | 위시 리스트 상품 삭제 | 회원의 위시 리스트에서 상품을 삭제한다.   |
| /api/wishes | GET    | 위시 리스트 상품 조회 | 회원의 위시 리스트에 있는 상품을 조회한다. |
| /api/wishes | PATCH | 위시 리스트 상품 수량 수정 | 회원의 위시 리스트에 있는 상품 수량을 수정한다. | 

✅ 구현 세부 사항

- 프로젝트 구조를 도메인 단위로 패키지를 구성하도록 변경했습니다.
- 기존 JWT 토큰에는 이메일, 비밀번호 정보만을 담았지만 사용자 식별을 위해 `member_id`를 포함했습니다.
- `HandlerMethodArgumentResolver`과 커스텀 어노테이션을 구현하여 컨트롤러에서 `member_id`를 사용할 수 있도록 했습니다.

<br><br>

테스트 코드까지 작성 후 코드 리뷰 요청드리고 싶었지만, 리뷰 요청이 늦어질 것 같아 먼저 PR 드립니다.

테스트 코드까지 작성해서 다시 리뷰 요청드리겠습니다.  감사합니다!

jinwoojwa committed:
refactor: 도메인 단위 아키텍처 적용 및 관련 import/경로 정리

jinwoojwa committed:
docs(README): 기능 요구 사항 정리

jinwoojwa committed:
feat: wishlist 테이블 DDL 추가

jinwoojwa committed:
feat(jwt): 토큰에 member_id claim 추가 및 관련 로직 수정

jinwoojwa committed:
feat(wishlist): 위시리스트 도메인 및 상품 추가 API 구현

- Wishlist 엔티티 추가
- WishlistRequestDto, WishlistResponseDto record 정의
- 위시리스트 상품 추가 API 컨트롤러, 서비스, 리포지토리 레이어 구현
- 상품이 이미 존재할 경우 수량 증가, 없을 경우 새로 추가
- JWT에서 memberId 추출하여 위시리스트에 연동

jinwoojwa committed:
refactor(auth): @LoginMember 어노테이션 기반 사용자 인증 리팩토링

- HttpServletRequest에서 직접 memberId 추출하던 방식 제거
- @CurrentUser 커스텀 어노테이션 및 ArgumentResolver 도입
- WebMvcConfigurer 통해 Resolver 등록

jinwoojwa committed:
위시 리스트 상품 조회 API 추가

jinwoojwa committed:
위시 리스트 상품 삭제 API 추가

jinwoojwa committed:
docs(README): 구현 상세 설명 작성

jinwoojwa committed:
위시 리스트 상품 수량 수정 API 추가

jinwoojwa committed:
refactor: 위시리스트 예외 커스텀 처리 및 RowMapper 중복 제거

- IllegalArgumentException → WishlistItemNotFoundException으로 변경
- wishlistResponseMapper을 도입하여 ResponseDto 매핑 코드 중복 제거

woodying commented on src/main/java/gift/common/annotation/LoginMember.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.lang.annotation.*;
+
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface LoginMember {
```
👍

woodying commented on src/main/java/gift/common/security/JwtAuthenticationFilter.java:
```diff
@@ -15,21 +17,41 @@ public JwtAuthenticationFilter(JwtUtil jwtUtil) {
+            return;
+        }
 
         String header = httpRequest.getHeader("Authorization");
```
사소하지만 `"Authorization"` 대신 `HttpHeaders.AUTHORIZATION` 을 사용할 수 있습니다 🙂

woodying commented on src/main/java/gift/common/security/JwtAuthenticationFilter.java:
```diff
@@ -15,21 +17,41 @@ public JwtAuthenticationFilter(JwtUtil jwtUtil) {
 
                 httpRequest.setAttribute("email", email);
                 httpRequest.setAttribute("role", role);
+                httpRequest.setAttribute("memberId", memberId);
```
이렇게 여러곳에서 사용하는 key 는 문자열 대신 const 로 정의해두면 좋을 것 같습니다 🙂

woodying commented on src/main/java/gift/wishlist/controller/WishlistController.java:
```diff
@@ -0,0 +1,67 @@
+    @PostMapping
+    public ResponseEntity<WishlistResponseDto> addProductToWishlist(
+            @RequestBody WishlistRequestDto requestDto,
+            @LoginMember Long memberId) {
```
👍

woodying reviewed (changes requested):
진우님 안녕하세요 🙂

도메인별로 패키지를 분리하셨네요~ 
구조를 보기에 훨씬 편한 것 같습니다! 👏

이번에도 역시 잘해주셔서 특별히 꼭 고쳐야할 부분은 없는 것 같은데요~
테스트코드만 추가해주시면 확인하고 approve 하겠습니다.
고생 많으셨어요 👍

jinwoojwa committed:
feat: 회원가입 및 로그인 API 테스트 추가

jinwoojwa committed:
fix: JWT 인증 필터에서 Authorization 헤더 상수화 및 속성 키 상수로 변경

- HttpHeaders.AUTHORIZATION 사용해 헤더 이름 상수화
- email, role, memberId 키를 상수로 분리하여 하드코딩 제거

jinwoojwa commented:
테스트 코드는 미션 진행하면서 계속 추가해나가겠습니다. 감사합니다.

woodying commented on src/test/java/gift/MemberApiConrollerTest.java:
```diff
@@ -0,0 +1,97 @@
+import org.springframework.http.ResponseEntity;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class MemberApiConrollerTest {
```
👍

woodying reviewed (approved):
진우님~ 두번째 미션 진행하시느라 고생 많으셨습니다 😊

너무 잘해주셔서 진우님 코드리뷰하기가 굉장히 편했던 것 같습니다.
진우님이라면 남은 미션들도 잘 수행해내실 것 같아요~!
남은 미션들도 파이팅입니다 🔥🔥

고생많으셨어요~!!
